### ISSUE TITLE
Interpretation

### ISSUE THREAD

**Issue by marciahon29** [ruiking04/COCA#22]
Hello,

I ran your models against my dataset. This dataset consists of 17 files with non-trivial anomalies (derived from the UCR anomalies).

I was expecting for COCA to perform the best overall. However, I have found that the top 5 performers in order are: OCSVM, DAGMM, Isolation Forest, LSTM-ED, and COCA - No Aug.

I am thinking that for my dataset, COCA is not necessary.

Please let me know your thoughts.

**Comment by ruiking04** [comment]
We found better hyperparameters on the 250 subsets of UCR to make the model perform well under the Affiliation F1 metric.

This may happen if only some subsets are found from UCR. On the other hand, the hyperparameters will be different depending on the data or evaluation indicators. So finding the best hyperparameters for deep models is a hard job.

Of course, you can also compare based on your ideas. Currently, deep model hyperparameters are still being further studied.

**Comment by marciahon29** [comment]
I would like to pursue tweeking with the hyperparameters for COCA. Please
could you help me get started? Are the following hyperparameters the ones
to change? How do you determine if a hyperparameter is better? Do you think
that if I change these variables that my COCA results will improve?

dataset UCR
input_channels 1
kernel_size 8
stride 1
final_out_channels 64
hidden_size 128
num_layers 3
project_channels 32
dropout 0.45
features_len 18
window_size 64
time_step 4
num_epoch 50
freeze_length_epoch 10
change_center_epoch 10
center_eps 0.1
omega1 1
omega2 0.1
beta1 0.9
beta2 0.99
lr 0.0003
drop_last FALSE
batch_size 512
nu 0.01
detect_nu 0.0005
threshold_determine one-anomaly
objective one-class
loss_type distance
scale_ratio 0.8
jitter_ratio 0.2

**Comment by ruiking04** [comment]
Some popular techniques for hyperparameter tuning in deep learning include grid search, random search, Bayesian optimization, and evolutionary algorithms. These methods involve systematically exploring the hyperparameter space to find the combination that results in the best model performance.

A popular tool for finding the best hyperparameters for deep learning models is the Hyperopt library, which provides a flexible and efficient implementation of Bayesian optimization for hyperparameter tuning. Other commonly used tools for hyperparameter tuning in deep learning include scikit-learn's GridSearchCV and RandomizedSearchCV, as well as libraries like Optuna and Ray Tune. These tools offer a range of hyperparameter optimization algorithms and can be used to efficiently search for the best hyperparameters for deep learning models.

You could adjust the five hyperparameters "window_size", "time_step",  "batch_size", "scale_ratio" and "jitter_ratio".

**Comment by ruiking04** [comment]
I think that deep learning is an experimental science. Most people generally see experimental results and then interpret them. This is why the deep model is called a black box model and is nicknamed "alchemy" by people. 

Training a good deep-learning model takes a lot of time and computing resources.

**Comment by marciahon29** [comment]
How did you fine tune the models? Did you use cross-validation for
testing/training?

**Comment by ruiking04** [comment]
traditional grid search

**Comment by marciahon29** [comment]
ok, I see. I will need to investigate how to perform this grid search.

with regards to the training/testing, did you do cross-validation with the
data?

**Comment by ruiking04** [comment]
We use "One round of cross-validation”

**Comment by marciahon29** [comment]
I'm sorry, I do not understand what you mean by "one round". I understand
cross-validation, but do not understand that.

**Comment by ruiking04** [comment]
One round of cross-validation involves partitioning a sample of data into complementary subsets, performing the analysis on one subset (called the training set), and validating the analysis on the other subset (called the validation set or testing set). To reduce variability, in most methods, multiple rounds of cross-validation are performed using different partitions, and the validation results are combined (e.g. averaged) over the rounds to give an estimate of the model's predictive performance.

Details can be found at: https://en.wikipedia.org/wiki/Cross-validation_(statistics)

**Comment by marciahon29** [comment]
OK, I understand. Just to let you know that multiple rounds of
cross-validation does not work on this data because it is sequential and
the anomalies are fixed.

**Comment by marciahon29** [comment]
Do you have any code I could work from that performs "traditional grid
search"?

I have determined that the COCA Hyperparameters to tweek are located at:
COCA/conf/coca/UCR_Configs.py . So I believe that all I need to do is to
change these variables and generate results. Is there an automatic way to
do this?

**Comment by marciahon29** [comment]
Did you tweek the hyperparameters for the models in baseline.py?

**Comment by ruiking04** [comment]
Yes, we modified some hyperparameters in baselines. However, we didn't modify the hyperparameters of those methods that are provided by Merlion themselves, because what they submitted was already the best ones.

**Comment by ruiking04** [comment]
I didn't use an automated search method, I created an Excel spreadsheet and manually logged the various hyperparameters ("window_size", "time_step", "batch_size", "scale_ratio" and "jitter_ratio") and the corresponding results. We ran almost a thousand experiments.

**Comment by marciahon29** [comment]
Thank you. I automated a Randomized Grid Search for 1600 iterations.
